src/hotspot/cpu/s390/gc/shared/barrierSetNMethod_s390.cpp

```diff
@@ -23,18 +23,108 @@
  */
 
 #include "precompiled.hpp"
+#include "asm/assembler.inline.hpp"
+#include "code/codeBlob.hpp"
+#include "code/nativeInst.hpp"
+#include "code/nmethod.hpp"
 #include "gc/shared/barrierSetNMethod.hpp"
 #include "utilities/debug.hpp"
 
+class NativeMethodBarrier: public NativeInstruction {
+  private:
+    static const int GUARD_INSTRUCTION_OFFSET = 3*6; // bytes
+
+    address get_barrier_start_address() const {
+      return NativeInstruction::addr_at(0);
+    }
+
+    address get_patchable_data_address() const {
+      address inst_addr = get_barrier_start_address() + GUARD_INSTRUCTION_OFFSET;
+
+      debug_only(Assembler::is_z_cfi(*((long*)inst_addr)));
+      return inst_addr + 2;
+    }
+
+  public:
+    static const int BARRIER_TOTAL_LENGTH = GUARD_INSTRUCTION_OFFSET + 2*6 + 2; // bytes
```

Comment:
GUARD_INSTRUCTION_OFFSET is the offset to the beginning of the patchable instruction. So, I believe it should be: 6 (cfi) + 6 (larl) + 2 (bcr).

It may be worth renaming 'GUARD_INSTRUCTION_OFFSET' as I feel it's a bit confusing.

Reply:
Ok, I thought bcr was 4 bytes. That may be wrong.